### PR TITLE
dafny: 2.3.0 -> 3.7.3

### PR DESCRIPTION
**Pull request by jwiegley**
###### Description of changes

* With minor fixes needed to boogie

This PR updates `dafny` to the latest release version 3.7.3, which is also compatible with the latest versions of Boogie and z3 that are found in nixpkgs.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by layus** [comment]
Thanks a lot for having a look. This looks very similar to what I reached on my side. Sadly, on my linux it fails with https://gist.github.com/layus/abe0ecd417bd3a3de4de44da3eff583b

Seems to be because of 

```
/build/dafny-3.7.3/Source/DafnyRuntime/DafnyRuntime.csproj(47,5): error MSB3073: The command "./gradlew -q build" exited with code 126.
```

Where it tries to run gradle to package a small java app.

**Review by layus — CHANGES_REQUESTED** [review]
Thanks a lot for having a look. This looks very similar to what I reached on my side. Sadly, on my linux it fails with https://gist.github.com/layus/abe0ecd417bd3a3de4de44da3eff583b

Seems to be because of 

```
/build/dafny-3.7.3/Source/DafnyRuntime/DafnyRuntime.csproj(47,5): error MSB3073: The command "./gradlew -q build" exited with code 126.
```

Where it tries to run gradle to package a small java app.

**Comment by jwiegley** [comment]
Thanks for trying it on Linux, later I'll get the build going in my VM and see what I can learn, unless you beat me to it.

**Comment by jwiegley** [comment]
Can confirm it happens here. Looks like `dotnet tool restore` wants access to the Internet during the build?

File: pkgs/top-level/dotnet-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/dotnet-packages.nix:157]:
```suggestion
        rm -f $out/bin/{System,Microsoft,NUnit3}.* $out/bin/*Tests
```

File: pkgs/top-level/dotnet-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/dotnet-packages.nix:179]: Please use fetchFromGitHub

File: pkgs/top-level/dotnet-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/dotnet-packages.nix:185]:
```suggestion
      ln -s ${z3} Binaries/z3
```

File: pkgs/top-level/dotnet-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/dotnet-packages.nix:188]:
```suggestion
    buildInputs = [ Boogie jdk11 ];
```

File: pkgs/top-level/dotnet-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/dotnet-packages.nix:207]:
```suggestion
      rm -f $out/bin/{coverlet,Microsoft,NUnit3,System}.*
      rm -f $out/bin/{ThirdPartyNotices.txt,XUnitExtensions}
```

File: pkgs/top-level/dotnet-packages.nix
Comment (jwiegley) [reply pkgs/top-level/dotnet-packages.nix:183]: With this change it would not evaluate.

File: pkgs/top-level/dotnet-packages.nix
Comment (jwiegley) [reply pkgs/top-level/dotnet-packages.nix:186]: Likewise.

**Comment by layus** [comment]
It's very strange indeed.

The dotnet restore call seems to fail because  cocor and dotnet-format are not part of the generated dependencies. This is why they cannot be restored without network access.

And then gradle also tries to access the network later on.

```
Package "cocor" failed to restore, due to Microsoft.DotNet.ToolPackage.ToolPackageException: The tool package could not be restored.
     at Microsoft.DotNet.Tools.Tool.Install.ProjectRestorer.Restore(FilePath project, PackageLocation packageLocation, String verbosity)
     at Microsoft.DotNet.ToolPackage.ToolPackageInstaller.InstallPackageToExternalManagedLocation(PackageLocation packageLocation, PackageId packageId, VersionRange versionRange, String targetFramework, String verbosity)
     at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.InstallPackages(ToolManifestPackage package, Nullable`1 configFile)
  Package "dotnet-format" failed to restore, due to Microsoft.DotNet.ToolPackage.ToolPackageException: The tool package could not be restored.
     at Microsoft.DotNet.Tools.Tool.Install.ProjectRestorer.Restore(FilePath project, PackageLocation packageLocation, String verbosity)
     at Microsoft.DotNet.ToolPackage.ToolPackageInstaller.InstallPackageToExternalManagedLocation(PackageLocation packageLocation, PackageId packageId, VersionRange versionRange, String targetFramework, String verbosity)
     at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.InstallPackages(ToolManifestPackage package, Nullable`1 configFile)
```

**Comment by jwiegley** [comment]
@layus Any ideas on what to look into to resolve that? I have almost zero experience with dotnet projects in the past ten years.

**Comment by layus** [comment]
Well, here are my findings:

:one: For the first issue (dotnet restore accessing the internet) the issue comes from `dotnet-tools.json` that dotnet restore tries to download but that are not picked by the dependency generation script for some reason. As there are only two of them, we could generate them manually.

All it took was to find this idea ;-) :

```diff
diff --git a/pkgs/development/dotnet-modules/dafny-deps.nix b/pkgs/development/dotnet-modules/dafny-deps.nix
index f2dd2b03661..35949673dd6 100644
--- a/pkgs/development/dotnet-modules/dafny-deps.nix
+++ b/pkgs/development/dotnet-modules/dafny-deps.nix
@@ -1,4 +1,8 @@
 { fetchNuGet }: [
+  # These two are added manually as they come from /dotnet-tools.json
+  (fetchNuGet { pname = "CocoR"; version = "2014.12.24"; sha256 = "0ps8h7aawkcc1910qnh13llzb01pvgsjmg862pxp0p4wca2dn7a2"; })
+  (fetchNuGet { pname = "dotnet-format"; version = "5.1.250801"; sha256 = "1i0icx2yyp9141rjb2a221a71fvsy0knrfyvv631vb56r8fnsywh"; })
+  # Generated stuff
   (fetchNuGet { pname = "Boogie.AbstractInterpretation"; version = "2.15.6"; sha256 = "0pdliiwc4185dm0skkafzz79njigyg5jq704mka928x51yrpl7mp"; })
   (fetchNuGet { pname = "Boogie.BaseTypes"; version = "2.15.6"; sha256 = "1mqibpc1d9vgi7lknys8l4jvl25h9n0h72msvjkdyr5n1821g00q"; })
   (fetchNuGet { pname = "Boogie.CodeContractsExtender"; version = "2.15.6"; sha256 = "1h6r1ak2c85gascfpi4k8cfgn2k95bd0racxg0m2mhcqnb9wc9rz"; })
```


:two: I think gradle is used to generate only a single executable jar file from a few internal file. We could patch it out and call java directly, or try to add gradle as a dependency, but then we have to look into how gradle packages are built within nixpkgs, and provide the same prefetched dependencies as with nuget. Or take the one you generated and add it to the build ? Or maybe extract it from the upstream binary ?

**Comment by layus** [comment]
01:00 am... Fixed in an ugly way @ https://github.com/layus/nixpkgs/commit/624c3df915e5c82cd81c1d94c243c7f7877d7ece

For :one: adding the dependency is not enough. It needs to be restored with a separate command. Also added the corresponding invocation to the fetcher script. This will trigger a rebuild and often require an update to all the dotnet packages.

For :two: I replaced the gradlew wrapper around gradle with a javac+jar invocation. The jar is so damn small and not used unless you compile dafny to Java. Dafny is usually compiled to C#, if compiled at all.

image

**Comment by jwiegley** [comment]
@layus Other than the postUnpack debugging change, I'd take it. Should I just pull it in as-is, or do you want to smooth it a bit more?

**Comment by layus** [comment]
We need a review from a maintainer of buildDotnetModule infra before merging a change there, and proper tests of the impact on all the other dotnet modules. Do you know who they are ?

**Comment by jwiegley** [comment]
Unfortunately I have no awareness of dotnet and its build infrastructure, and especially not as it relates to nixpkgs. I only waded into these waters because I have a need to get dafny updated....

**Comment by layus** [comment]
So you do need dafny after all :-).

**Comment by layus** [comment]
Hi @IvarWithoutBones,

could you have a look at https://github.com/layus/nixpkgs/commit/624c3df915e5c82cd81c1d94c243c7f7877d7ece to see if it makes sense for you ? It adds dotnet tools to the prefetched sources with the other dependencies. Does it seems like a good idea to you ?

Thanks for your insights.

**Comment by layus** [comment]
Okay, just nix-review'd my patch, and it breaks at least 13 packages, because it tries to restore tools that are not in the generated dependencies. I am not sure which path is best. Updating all the 13 packages to support the unified workflow that includes tools in the dependencies, or add a flag that toggles the usage of tools.

And, if we go the flag way, should it be on by default, or off. Meaning, do I need to add the flag to dafny to make it work, or to all the 13 packages to hint that I did not take the time to fix them ?

**Comment by layus** [comment]
@jwiegley  Sadly I cannot edit this PR, but please merge https://github.com/jwiegley/nixpkgs/pull/44 and the changes will appear here.

File: pkgs/build-support/dotnet/build-dotnet-module/default.nix
Comment (mdarocha) [inline pkgs/build-support/dotnet/build-dotnet-module/default.nix:61]: Please try with https://github.com/NixOS/nixpkgs/pull/185021, the changes in that PR should fix this flag for dotnet <6. Then, this `null` logic can be simplified.

**Review by mdarocha — CHANGES_REQUESTED** [review]

**Review by layus — APPROVED** [review]
Seems to have matured to a stable place. I am obviously biased since I wrote most of it ;-)

@jwiegley can you give it a last try ? Maybe check that it works, and that there are no stray file left in the output path ?

**Review by mdarocha — APPROVED** [review]

**Comment by IvarWithoutBones** [comment]
Hi! Thanks for taking the time to include this fix in `buildDotnetModule`, fixing this for future packages definitely seems like the best move.

Seeing how it breaks packages that do not include a `dotnet tool`, I think it would be best to guard it behind a flag, something like `restoreDotnetTool` perhaps? I think it makes the most sense for it to be disabled by default, since (as far as i know) it's not extremely common. 

Also, looking at the commit in this PR making the addition to `buildDotnetModule` also includes lockfile regeneration of unrelated packages, I think those should be done in a seperate commit.

File: pkgs/build-support/dotnet/build-dotnet-module/default.nix
Comment (IvarWithoutBones) [inline pkgs/build-support/dotnet/build-dotnet-module/default.nix:139]: Is there a way to specify the location to restore to? If not, we should probably symlink instead of copying below this.

File: pkgs/top-level/dotnet-packages.nix
Comment (IvarWithoutBones) [inline pkgs/top-level/dotnet-packages.nix:173]: I would prefer for this to be done in a seperate file instead of `dotnetPackages.nix`, it makes the package easier to discover IMO.


This file is mainly used for legacy `buildDotnetPackage` derivations, because of this I also believe splitting this package away would be best.

**Comment by superherointj** [comment]
@ofborg eval

**Comment by jwiegley** [comment]
@layus Since I've been a bit busy recently, if you want to fork my branch and start a new PR, I'd be happy to close this and move the discussion there. I still have to give this current version a try.

**Comment by jwiegley** [comment]
GM> So you do need dafny after all :-).

Haha, yes. My main target is Boogie, but a friend who works on Dafny suggested I use it to draft what I intend to convert, and crib from the generated Boogie rather than figuring out how to do everything from scratch. Working well so far!

**Comment by layus** [comment]
Hi Ivv,

I had an initial implementation with a guarding flag, but I figured out that 1) it seems to work for all packages even if there is no tools to restore, and no tools-config-file-thing and 2) the overhead is pretty small anyway.

Given the small number of dotnet modules, I decided to take the time to update them all. Now there is no flag to discover for packagers, so that's less complexity at a small cost.

You mention that it breaks packages. Can you give me an example ? You also say that it updates the lockfile. Do you also have an example ? Some packages have seen their dependencies updated in this PR. I was under the impression that it was because of previous version bumps where the fetcher script was not run. Maybe I am wrong.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
please rebase

**Review by mdarocha — APPROVED** [review]

**Comment by whonore** [comment]
What, aside from a rebase, still needs to be done to push this over the finish line? And maybe while we're at it could we bump the version to 3.13.1 or even 4.0.0?

**Comment by jwiegley** [comment]
I'm not sure. Since I'm not using Dafny at work anymore, I don't have the gumption to finish this, so happen to let anyone else take over the work if they'd like to clone my branch.

**Comment by SuperSandro2000** [comment]
Since you are not planning to work on this, I am going to close it.

**Pull request closed without merging**
